packages/substrate-wasm/src/index.ts

```diff
@@ -0,0 +1,521 @@
+
+interface SpecDef {
+  messages: {
+    label: string;
+    name: string[] | string;
+    selector: string;
+  }[];
+
+  events: {
+    args: any[];
+    docs: string[];
+    label: string;
+  }[];
+}
+
+interface JSONAbi {
+  source: {
+    compiler: string;
+    hash: string;
+    language: string;
+    wasm: string;
+  };
+  spec: SpecDef;
+  V1: {
+    spec: SpecDef;
+  };
+  V2: {
+    spec: SpecDef;
+  };
+  V3: {
+    spec: SpecDef;
+  };
+  V4: {
+    spec: SpecDef;
+  };
+}
+
+export interface Result extends ReadonlyArray<any> {
+  readonly [key: string]: any;
+}
+
+export type WasmDatasource = SubstrateCustomDatasource<
+  'substrate/Wasm',
+  SubstrateNetworkFilter,
+  SubstrateMapping<SubstrateCustomHandler>,
+  WasmProcessorOptions
+>;
+
+export interface WasmEventFilter extends SubstrateNetworkFilter {
+  from?: string;
+  contract?: string;
+  identifier?: IdentifierFilter;
+}
+
+export interface WasmCallFilter extends SubstrateNetworkFilter {
+  // dest?: string; //Filter in processor option contract
+  from?: string;
+  selector?: string; //To u8a
+  method?: string; // label
+}
+
+export type ContractEmittedResult = [AccountId, Bytes];
+
+export type ContractCallArgs = [Address, BalanceOf, Weight, Option<Compact<u128>>, Bytes];
+
+// TODO add valid_until, access_list
+export interface WasmEvent<T extends Result = Result> {
+  from: string;
+  contract: AccountId;
+  eventIndex: number;
+  identifier?: string | undefined;
+  args?: T | undefined;
+  transactionHash: string;
+  blockNumber: number;
+  blockHash: string;
+  timestamp: Date;
+}
+export interface WasmCall<T extends Result = Result> {
+  from: Address;
+  dest: Address;
+  value: BalanceOf;
+  gasLimit: Weight;
+  storageDepositLimit: Option<Compact<u128>>;
+  data: {args: T; message: AbiMessage} | string;
+  selector: string;
+  success: boolean;
+  hash: string;
+  blockNumber: number;
+  blockHash: string;
+  timestamp: Date;
+}
+
+class WasmProcessorOptions {
+  @IsOptional()
+  abi?: string;
+  @IsOptional()
+  contract?: string;
+}
+
+class WasmEventFilterImpl implements WasmEventFilter {
+  @IsOptional()
+  @IsString()
+  from?: string;
+  @IsOptional()
+  @IsString()
+  contract?: string;
+  @IsOptional()
+  // @Validate(TopicFilterValidator, {each: true}) //TODO
+  identifier?: IdentifierFilter;
+}
+
+class WasmCallFilterImpl implements WasmCallFilter {
+  @IsOptional()
+  @IsString()
+  dest?: string;
+  @IsString()
+  @IsOptional()
+  selector?: string;
+  @IsString()
+  @IsOptional()
+  method?: string; // label
+}
+
+const dsAssets: Record<string, string> = {};
+// used abi name with its ABI object will be stored
+const contractAbis: Record<string, Abi> = {};
+//only single event will be stored, reset if encounter new event
+let decodedEvent: Record<string, DecodedEvent> = {};
+//only single message will be stored, reset if encounter new message
+let decodedMessage: Record<string, DecodedMessage> = {};
```

Comment:
Do you think its an issue that items never get removed from these and would continuously grow?